packages/plugin-hardhat/src/utils/deploy-impl.ts

```diff
@@ -1,66 +1,154 @@
+interface DeployedBeaconImpl {
+  impl: string;
+}
+
+async function getDeployData(hre: HardhatRuntimeEnvironment, ImplFactory: ContractFactory, opts: Options) {
   const { provider } = hre.network;
   const validations = await readValidations(hre);
   const unlinkedBytecode = getUnlinkedBytecode(validations, ImplFactory.bytecode);
   const encodedArgs = ImplFactory.interface.encodeDeploy(opts.constructorArgs);
   const version = getVersion(unlinkedBytecode, ImplFactory.bytecode, encodedArgs);
   const layout = getStorageLayout(validations, version);
+  const fullOpts = withDefaults(opts);
+  return { provider, validations, unlinkedBytecode, encodedArgs, version, layout, fullOpts };
+}
 
-  if (opts.kind === undefined) {
-    opts.kind = inferProxyKind(validations, version);
-  }
+export async function deployProxyImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  proxyAddress?: string,
+): Promise<DeployedProxyImpl> {
+  const deployData = await getDeployData(hre, ImplFactory, opts);
+  await processProxyKind();
 
+  let currentImplAddress;
   if (proxyAddress !== undefined) {
-    await setProxyKind(provider, proxyAddress, opts);
+    // upgrade scenario
+    currentImplAddress = await getImplementationAddress(deployData.provider, proxyAddress);
   }
 
-  const fullOpts = withDefaults(opts);
+  return deployImpl(deployData, ImplFactory, opts, currentImplAddress);
 
-  assertUpgradeSafe(validations, version, fullOpts);
+  async function processProxyKind() {
+    if (opts.kind === undefined) {
+      if (proxyAddress !== undefined && (await isBeaconProxy(deployData.provider, proxyAddress))) {
+        opts.kind = 'beacon';
+      } else {
+        opts.kind = inferProxyKind(deployData.validations, deployData.version);
+      }
+    }
 
-  if (proxyAddress !== undefined) {
-    const manifest = await Manifest.forNetwork(provider);
-    const currentImplAddress = await getImplementationAddress(provider, proxyAddress);
-    const currentLayout = await getStorageLayoutForAddress(manifest, validations, currentImplAddress);
-    assertStorageUpgradeSafe(currentLayout, layout, fullOpts);
+    if (proxyAddress !== undefined) {
+      await setProxyKind(deployData.provider, proxyAddress, opts);
+    }
+
+    if (opts.kind === 'beacon') {
+      throw new BeaconProxyUnsupportedError();
+    }
+  }
+}
+
+export async function deployBeaconImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  beaconAddress?: string,
+): Promise<DeployedBeaconImpl> {
+  const deployData = await getDeployData(hre, ImplFactory, opts);
+
+  let currentImplAddress;
+  if (beaconAddress !== undefined) {
+    // upgrade scenario
+    currentImplAddress = await getBeaconImplementationAddress(beaconAddress);
+  }
+  return deployImpl(deployData, ImplFactory, opts, currentImplAddress);
+
+  async function getBeaconImplementationAddress(beaconAddress: string): Promise<string> {
```

Comment:
This one would also be a good fit for core, but not in the way it's written now. We would need to change it so it doesn't use Ethers.js. We can just manually craft the `implementation()` function call and use it with `eth_call` through JSON RPC.